vision/odlc_contour_filtering.py

```diff
@@ -0,0 +1,407 @@
+
+    parameters
+    ----------
+    contour : contour_type
+        The individual contour to be evaluated (as returned from cv2.findContours)
+    dims : tuple[int, int]
+        The dimensions of the image the contour is from
+    test_area_ratio : float
+        The minimum acceptable ratio of image area to contour bounding box area
+
+    returns
+    -------
+    returns true if the bounding box area is less than the image area by a factor of
+    test_area_ratio or more
+    """
+    bounding_box: bound_box_type = cv2.boundingRect(contour)
+
+    box_area: float = abs(bounding_box[2] - bounding_box[0]) * abs(
+        bounding_box[3] - bounding_box[1]
+    )
+    img_area: float = dims[0] * dims[1]
+
+    return box_area * test_area_ratio <= img_area
+
+
+def test_spikiness(contour: contour_type) -> bool:
+    """
+    Checks whether the contour has some points that are much farther away from the center of mass
+    than the rest of the points
+    basically real shapes will not have random points that are really far away from the rest
+    but a weird patch of grass or something without well defined edges will have crazy points
+    all over the place
+
+    parameters
+    ----------
+    contour : contour_type
+        The individual contour to be evaluated (as returned from cv2.findContours)
+
+    returns
+    -------
+    True unless there is a statistical outlier point that is uniquely far away (or close ig)
+    """
+    moments: dict[str, float] = cv2.moments(contour)
+    # com is Center of Mass, com = (y_coord, x_coord)
+    com: NDArray[Shape["2"], Float32] = np.array(
+        ((moments["m01"] / moments["m00"]), (moments["m10"] / moments["m00"])), dtype=np.float64
+    )
+
+    dists_com: NDArray[Shape["*"], Float32] = np.ndarray(contour.shape[0], Float32)
+    for i in range(contour.shape[0]):
+        dists_com[i] = cv2.norm(contour[i] - com)
+
+    dists_median: float = np.median(dists_com)
+    dists_outlier_range: float = 1.5 * (np.quantile(dists_com, 0.75) - np.quantile(dists_com, 0.25))
+
+    if np.all(dists_com < dists_median + dists_outlier_range):
+        return True
+    return False
+
+
+def _min_common_bounding_box(contours: tuple[contour_type, ...]) -> bound_box_type:
+    """
+    takes a set of contours and returns the smallest bounding
+    box that encloses all of them
+
+    parameters
+    ----------
+    contours : tuple[contour_type]
+        Tuple of contours as formatted in cv2.findContour to find the minimum common bounding box
+
+    returns
+    -------
+    the smallest bounding box that encompases all of the given contours
+    """
+
+    boxes: tuple[bound_box_type, ...] = ()
+    for contour in contours:
+        boxes += (cv2.boundingRect(contour),)
+
+    min_box: bound_box_type = (
+        min(box[0] for box in boxes),
+        min(box[1] for box in boxes),
+        max(box[2] for box in boxes),
+        max(box[3] for box in boxes),
+    )
+    return min_box
```

Comment:
The `BoundingBox` type under vision > common has some functions that may be useful for some of the stuff you are doing here.